File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] seenChars = new int[26];
        int left = 0;
        int max = 0;
        int maxFrequency = 0;

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);
            seenChars[currentChar - 'A']++;
            maxFrequency = Math.max(maxFrequency, seenChars[currentChar - 'A']);

            while (right - left + 1- maxFrequency > k) {
                char leftChar = s.charAt(left);
                seenChars[leftChar - 'A']--;
                left++;
            }

            max = Math.max(max, right - left + 1);
        }


        return max;
    }
}

File: Data Structures & Algorithms/is-palindrome/submission-4.kt
class Solution {
    fun isPalindrome(s: String): Boolean {
        var left = 0
        var right = s.length - 1

        while (left < right) {
            if (!s[left].isLetterOrDigit()) {
                left++
            } else if (!s[right].isLetterOrDigit()) {
                right--
            } else if (s[left].lowercaseChar() != s[right].lowercaseChar()) {
                return false
            } else {
                left++
                right--
            }
        }
        return true
    }
}
